pkg/controller/drivelvgcontroller/drivelvgcontroller.go

```diff
@@ -0,0 +1,365 @@
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	api "github.com/dell/csi-baremetal/api/generated/v1"
+	apiV1 "github.com/dell/csi-baremetal/api/v1"
+	"github.com/dell/csi-baremetal/api/v1/drivecrd"
+	"github.com/dell/csi-baremetal/api/v1/lvgcrd"
+	"github.com/dell/csi-baremetal/pkg/base"
+	"github.com/dell/csi-baremetal/pkg/base/capacityplanner"
+	errTypes "github.com/dell/csi-baremetal/pkg/base/error"
+	"github.com/dell/csi-baremetal/pkg/base/k8s"
+	"github.com/dell/csi-baremetal/pkg/base/util"
+	metricsC "github.com/dell/csi-baremetal/pkg/metrics/common"
+)
+
+// RequeueDriveTime is time between drives reconciliation
+const RequeueDriveTime = time.Second * 30
+
+// Controller reconciles drive custom resource
+type Controller struct {
+	client   *k8s.KubeClient
+	crHelper *k8s.CRHelper
+	// CRHelper instance which reads from cache
+	cachedCrHelper *k8s.CRHelper
+	log            *logrus.Entry
+}
+
+// NewDriveController creates new instance of Controller structure
+// Receives an instance of base.KubeClient and logrus logger
+// Returns an instance of Controller
+func NewDriveController(client *k8s.KubeClient, k8sCache k8s.CRReader, log *logrus.Logger) *Controller {
+	return &Controller{
+		client:         client,
+		crHelper:       k8s.NewCRHelper(client, log),
+		cachedCrHelper: k8s.NewCRHelper(client, log).SetReader(k8sCache),
+		log:            log.WithField("component", "Controller"),
+	}
+}
+
+// SetupWithManager registers Controller to ControllerManager
+func (d *Controller) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&drivecrd.Drive{}).
+		Watches(&source.Kind{Type: &lvgcrd.LogicalVolumeGroup{}}, &handler.EnqueueRequestForObject{}).
+		WithEventFilter(predicate.Funcs{
+			DeleteFunc: func(e event.DeleteEvent) bool {
+				return false
+			},
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				return d.filterUpdateEvent(e.ObjectOld, e.ObjectNew)
+			},
+		}).
+		Complete(d)
+}
+
+// Reconcile reconciles Drive custom resources
+func (d *Controller) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	defer metricsC.ReconcileDuration.EvaluateDurationForType("csicontroller_drive_controller")()
+	resourceName := req.Name
+	ctx, cancelFn := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancelFn()
+
+	log := d.log.WithFields(logrus.Fields{"method": "Reconcile", "name": resourceName})
+
+	drive := &drivecrd.Drive{}
+	if err := d.client.ReadCR(ctx, resourceName, "", drive); err == nil {
+		return d.reconcileDrive(ctx, drive)
+	}
+	log.Warnf("Failed to read Drive %s CR, try to read LVG CR", resourceName)
```

Comment:
Please check for not found error and print warning when all resources not found.

Reply:
Fixed